### ISSUE TITLE
Application "upgrade" solution

### ISSUE THREAD

**Issue by rayzer42** [metacurrency/holochain#433]
The rough idea Eric and I discussed is two new core functions: upgrade() and validateUpgrade().

upgrade() would update the DNA in the local chain and install new source files into the file system, and then distribute the operation into the peer network. (maybe not in that order)

validateUpgrade() would allow a node receiving an upgrade() from the peer network to perform the checks the app developer built in before doing the same.

The big question is how does an upgrade get triggered?

In my mind it is critical that a simple, "self-upgrading" solution is built into our sample apps and code generators. When newbie app devs put out their first app, we don't want them to find out that after a week of running and collecting meaningful data into the chains, they have no way to even fix a bug without publishing a whole new app and holochain. Even forcing the developer to build UI that has a user or users trigger an upgrade would be asking too much as it could get easily get missed by a newbie.

I assigned this to beta since, as a developer, I wouldn't even bother to start writing an app until I knew how to upgrade my app.

There is a reason Ethereum is 2 years old, and there is not a single useful app out in the world yet. And the inability to "upgrade" a dumb contract is a major reason for that, IMHO.

Arthur, perhaps you can poopoo all this and express a more elegant and/or simpler proposal.

**Comment by rayzer42** [comment]
Actually, maybe the functions should be fork() and validateFork() !!! In theory, any node running an app could fork() itself anytime, and ask others to join it! If conflict arises, you fork(). Fork() you, app!

**Comment by rayzer42** [comment]
related to #436

**Comment by artbrock** [comment]
Yeah... #436 was my suggestion for keeping it simple (ish) and flexible for upgrading and forking. I think there's something clean about the new version being independent (not created through a forking function) but then the link/authorization/migration happening by publishing the forking-pointer on the old app. Which can do whatever validation, including governance related checks, in accepting that entry.

**Comment by christopherreay** [comment]
given the security restrictions, the only difference I can see between"forking" and newAppOldApp is how "old" data is accessed. how many holochains have to be running on each machine *after* the upgrade is complete?

also there's this half on new half on old which I didn't see addressed in this conversation
